internal/dag/gatewayapi_processor.go

```diff
@@ -1310,6 +1393,208 @@ func (p *GatewayAPIProcessor) computeHTTPRoute(route *gatewayapi_v1beta1.HTTPRou
 	return programmed, hosts
 }
 
+func (p *GatewayAPIProcessor) computeGRPCRoute(route *gatewayapi_v1alpha2.GRPCRoute, routeAccessor *status.RouteParentStatusUpdate, listener *listenerInfo) (bool, sets.Set[string]) {
+	hosts, errs := p.computeHosts(route.Spec.Hostnames, gatewayapi.HostnameDeref(listener.listener.Hostname))
+	for _, err := range errs {
+		// The Gateway API spec does not indicate what to do if syntactically
+		// invalid hostnames make it through, we're using our best judgment here.
+		// Theoretically these should be prevented by the combination of kubebuilder
+		// and admission webhook validations.
+		routeAccessor.AddCondition(gatewayapi_v1beta1.RouteConditionResolvedRefs, metav1.ConditionFalse, status.ReasonDegraded, err.Error())
+	}
+
+	// If there were no intersections between the listener hostname and the
+	// route hostnames, the route is not programmed for this listener.
+	if len(hosts) == 0 {
+		return false, nil
+	}
+
+	var programmed bool
+	for ruleIndex, rule := range route.Spec.Rules {
+		// Get match conditions for the rule.
+		var matchconditions []*matchConditions
+		for _, match := range rule.Matches {
```

Comment:
I just filed https://github.com/kubernetes-sigs/gateway-api/issues/1751 upstream, but we'll likely want to add code here to handle a rule with 0 matches, which should result in matching all requests. This can be expressed as a prefix match on `/`.

Reply:
Do we need this? For HTTPRoute, I don't see such kind of implementation for the case of a rule with 0 matches. But for a rule with matches but no method match, we need to add a prefix match on /. I will handle the latter case.

Reply:
I think we do need this in order to match the spec:

> If no matches are specified, the implementation MUST match every gRPC request.

(ref. https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRouteRule)

It looks like HTTPRoute has similar language, but the API defaulting means that in practice Contour will probably never see an HTTPRouteRule with 0 matches:

> If no matches are specified, the default is a prefix path match on “/”, which has the effect of matching every HTTP request.

(ref. https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteRule)

API defaulting: https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/httproute_types.go#L179

Reply:
Thank you for pointing this out. I will make changes accordingly.